### PR TITLE
add @property to sklearn interface

### PR DESCRIPTION
**Pull request by wxchan**
related to #154 
In sklearn, feature_importance_ is an attribute rather than a method. It will be dumped and loaded by pickle/joblib.

Also clean some codes in fit().
1. params["objective"] already equals to self.objective with self.get_params(), make sure it's not a callable function
2. eval_metric can have other options(such as set), let params_to_dict check it
3. the following code seems not necessary, what's it for
```
for val in evals_result.items(): 
  evals_result_key = list(val[1].keys())[0]
  evals_result[val[0]][evals_result_key] = val[1][evals_result_key]
```

### CODE REVIEW

File: python-package/lightgbm/sklearn.py [hunk 1]
```diff
@@ -451,34 +426,26 @@ def apply(self, X, num_iteration=0):
...
-        else:
-            raise LightGBMError('No results found.')
-
-        return evals_result
-
-    def feature_importance(self):
-        """
-        Feature importances
+    @property
+    def booster_(self):
```
Comment (henry0312) [inline python-package/lightgbm/sklearn.py]: is Booster really property?

Comment (wxchan) [inline python-package/lightgbm/sklearn.py]: sklearn does not have similar method, it's more like estimators_ I think.

I misunderstood this method before. The main usage of this method is to check _Booster is not None before we access to it, so it's more like a property method does. Do you think it's reasonable?

File: python-package/lightgbm/sklearn.py [hunk 2]
```diff
@@ -558,18 +526,12 @@ def fit(self, X, y,
...
-                                            num_iteration=num_iteration)
-        if len(class_probs.shape) > 1:
-            column_indexes = np.argmax(class_probs, axis=1)
-        else:
-            column_indexes = np.repeat(0, class_probs.shape[0])
-            column_indexes[class_probs > 0.5] = 1
-        return self._le.inverse_transform(column_indexes)
+    def predict(self, X, raw_score=False, num_iteration=0):
+        class_probs = self.predict_proba(X, raw_score, num_iteration)
+        class_index = np.argmax(class_probs, axis=1)
```
Comment (guolinke) [inline python-package/lightgbm/sklearn.py]: can this work for binary classification ?

File: python-package/lightgbm/sklearn.py (hunk 1 above)
Comment (guolinke) [inline python-package/lightgbm/sklearn.py]: why not ```booster```?

Comment (henry0312) [inline python-package/lightgbm/sklearn.py]: It's difficult to decide to use property or method.
The key is whether we allow users to set `self._Booster`.
I guess `self._Booster` is a **private** member (because of the prefix of `_` ), therefore it's better not to provide any getter, setter, and deleter (i.e. have one as property).
(If it is a property, we may access it like `gbm.booster_ = another_Booster`)

Moreover, the sklearn interface of LightGBM is very similar to that of XGBoost, so if we don't have any strong reasons, it may be bettor to follow the way of XGBoost.
(cf. https://github.com/dmlc/xgboost/blob/master/python-package/xgboost/sklearn.py#L153)

File: python-package/lightgbm/sklearn.py [hunk 3]
```diff
@@ -349,19 +337,14 @@ def fit(self, X, y,
             params['num_class'] = self.n_classes_
         if hasattr(self, 'eval_at'):
             params['ndcg_eval_at'] = self.eval_at
-
         if self.fobj:
-            params["objective"] = "None"
-        else:
-            params["objective"] = self.objective
+            params.pop('objective', None)
```
Comment (guolinke) [inline python-package/lightgbm/sklearn.py]: if pop ```objective```, the default objective in c_api will change to ```regression```.

**Comment by guolinke** [comment]
@wxchan 
for 3: 
I think it groups the eval result from ```data, eval_metric``` to ```data, iteration```

File: python-package/lightgbm/sklearn.py (hunk 3 above)
Comment (wxchan) [inline python-package/lightgbm/sklearn.py]: is there a `objective=None` option? ok, I will make it `params["objective"] = "None"`.

File: python-package/lightgbm/sklearn.py (hunk 2 above)
Comment (wxchan) [inline python-package/lightgbm/sklearn.py]: it works, the predicted label = argmax of predicted proba

but np.argmax only returns first argmax if there are several max numbers, what we should return under this situation?

File: python-package/lightgbm/sklearn.py (hunk 3 above)
Comment (guolinke) [inline python-package/lightgbm/sklearn.py]: @wxchan, the factory method of objective will return nullptr if passed string is not pre-defined. So we can use ```"None"``` to create a null objective.

File: python-package/lightgbm/sklearn.py (hunk 2 above)
Comment (guolinke) [inline python-package/lightgbm/sklearn.py]: I think return of first max is OK. 
And lightgbm also use first max as predict class for multi-classification metrics.

**Comment by wxchan** [comment]
@guolinke do we need to group from data, eval_metric to data, iteration? xgboost only has one eval_metric, it might not work with same code.

File: python-package/lightgbm/sklearn.py (hunk 1 above)
Comment (wxchan) [inline python-package/lightgbm/sklearn.py]: @guolinke no, just sklearn names attributes with an underline suffix, see http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.GradientBoostingRegressor.html#sklearn.ensemble.GradientBoostingRegressor

@henry0312 you cannot set property either. `@booster_.setter` is a setter. self._Booster should be private but it won't be in python, so `gbm._Booster = another_Booster` won't raise Exception but `gbm.booster_ = another_Booster` will.
I think xgboost may not consider this question in their early stage, and it will be painful for them to change api now.

File: python-package/lightgbm/sklearn.py (hunk 3 above)
Comment (wxchan) [inline python-package/lightgbm/sklearn.py]: @guolinke ok, we will change it and add a comment here.

File: python-package/lightgbm/sklearn.py (hunk 2 above)
Comment (guolinke) [inline python-package/lightgbm/sklearn.py]: However, for the binary classification, i am not sure about the ``prob==0.5`` belongs to negative or positive. 
current code is positive.
https://github.com/Microsoft/LightGBM/blob/master/src/metric/binary_metric.hpp#L129-L135

But use argmax, it seems will convert to negative class.

which one is better?

File: python-package/lightgbm/sklearn.py (hunk 1 above)
Comment (henry0312) [inline python-package/lightgbm/sklearn.py]: Hum, you may be right...
I'll follow your idea that it's better to treat _Booster as a property :)

**Comment by wxchan** [comment]
@guolinke it's kinda like train_score_, evals_result is evals_score_.

Even if we don't delete it, we need to rewrite it: it doesn't do what we want, and it's super unreadable.

File: python-package/lightgbm/sklearn.py (hunk 2 above)
Comment (wxchan) [inline python-package/lightgbm/sklearn.py]: If there isn't a standard rule, both are fine. 0.5 is a bad prediction anyway.

Comment (wxchan) [inline python-package/lightgbm/sklearn.py]: If there isn't a standard rule, both are fine. 0.5 is a bad prediction anyway. Actually the old version 0.5 -> 0 as well.

Comment (guolinke) [inline python-package/lightgbm/sklearn.py]: @wxchan , I think you can change binary_metric.hpp to let it to ```0.5 -> 0``` as well.

**Comment by henry0312** [comment]
@wxchan can you leave `booster()` and `feature_importance()`? I don't think they are needed to be removed.

```python
def booster(self):
    return self.booster_

def feature_importance(self):
    return self.feature_importance_
```

**Comment by wxchan** [comment]
@henry0312 yes, but I will add `@deprecated` to them.

btw, look like this pr has nothing to do with #154 . The bug is caused by dump_model(), somehow be fixed somewhere.

**Comment by henry0312** [comment]
I think this PR should be necessary for https://github.com/Microsoft/LightGBM/issues/154#issuecomment-269908741 👍

**Comment by henry0312** [comment]
Oh, it's true that this PR won't fix https://github.com/Microsoft/LightGBM/issues/154#issuecomment-269908741 😢
What's wrong...?

**Comment by guolinke** [comment]
@wxchan  it seems break the test:
https://travis-ci.org/Microsoft/LightGBM/jobs/188203455#L732-L735

**Comment by wxchan** [comment]
it's the same bug with #156

**Comment by guolinke** [comment]
@wxchan Is this finished?

**Comment by wxchan** [comment]
@guolinke yes, finished.
